Kernel priorities

declanvk opened the pull request:
Change kernel execution model. Instead of offloading execution work to be picked from a queue, this will run each request thread, including continuous async polling, until "completion".

Hopefully use of `asyncio.sleep(0)` will be sufficient to show interleaving execution.

declanvk committed:
Change event loop to prevent too strict ordering through queue

The new execution loop will schedule its next iteration upon receiving from the queue, instead of polling in a single function. This new style allows async generators to schedule iterations without falling into an predefined order.

declanvk committed:
Add async lock based on dependency trackers

Also add testing for dependency tracker/lock, with documentation in README

ellisonbg commented:
Hi Declan, can you go into more depth about why this is needed? Also, are
you using actual threads?

On Fri, Sep 7, 2018 at 12:00 AM Declan Kelly <notifications@github.com>
wrote:

> Change kernel execution model. Instead of offloading execution work to be
> picked from a queue, this will run each request thread, including
> continuous async polling, until "completion".
>
> Hopefully use of asyncio.sleep(0) will be sufficient to show interleaving
> execution.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/jupytercalpoly/reactivepy/pull/26
> Commit Summary
>
>    - Change event loop to prevent too strict ordering through queue
>    - Add async lock based on dependency trackers
>
> File Changes
>
>    - *M* README.md
>    <https://github.com/jupytercalpoly/reactivepy/pull/26/files#diff-0>
>    (16)
>    - *M* reactivepy/dependencies.py
>    <https://github.com/jupytercalpoly/reactivepy/pull/26/files#diff-1>
>    (124)
>    - *M* reactivepy/kernel.py
>    <https://github.com/jupytercalpoly/reactivepy/pull/26/files#diff-2>
>    (187)
>    - *A* test/__init__.py
>    <https://github.com/jupytercalpoly/reactivepy/pull/26/files#diff-3>
>    (0)
>    - *A* test/test_dependencies.py
>    <https://github.com/jupytercalpoly/reactivepy/pull/26/files#diff-4>
>    (202)
>
> Patch Links:
>
>    - https://github.com/jupytercalpoly/reactivepy/pull/26.patch
>    - https://github.com/jupytercalpoly/reactivepy/pull/26.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupytercalpoly/reactivepy/pull/26>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABr0O_f4qO2jS7kASx3SMLppaLa-Oe4ks5uYhl3gaJpZM4WeSLQ>
> .
>


-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

declanvk commented:
@ellisonbg The description doesn't really say that much, sorry about that.

This PR will remove the use of a queue for execution. Previously, any type of execution, from a request or from an iterator, was put into a queue and then polled off one by one.

The issue with this is that as part of the execution step the next iteration will be added to the back of the queue. Then the execution order for two iterators will always go `a b a b a b` even if `a` and `b` complete in 10 seconds and 1 second respectively.

The solution to this is to not use the queue and instead complete the execution as part of the request, after the step of extracting variables and updating the dependency graph.

declanvk committed:
Remove dependency-based locking

declanvk committed:
Add pylint config file

declanvk committed:
Add comments and typings, remove unused exceptions

declanvk committed:
Remove use of locking/queue for async iterations

May not be as robust in cornercases, but performs better in many common scenarios. For example, now dependent async iterators can restart one another without erroring out and will output without starving faster iterators.

declanvk committed:
Merge branch 'master' into kernel-priorities

declanvk committed:
Remove use of local namespace

declanvk committed:
Removed extra logging statements

declanvk committed:
Add CitiBike live graph example

declanvk commented:
@ellisonbg This PR is ready to go, if you have a moment to look at it. I've also added an example the creates a live graph of the New York CitiBike availability.
